xgboost_ray/data_sources/partitioned.py

```diff
@@ -0,0 +1,98 @@
+from typing import Any, Optional, Sequence, Dict, Union, Tuple
+
+from collections import defaultdict
+import pandas as pd
+import numpy as np
+
+import ray
+from ray import ObjectRef
+from ray.actor import ActorHandle
+
+from xgboost_ray.data_sources._distributed import \
+    assign_partitions_to_actors, get_actor_rank_ips
+from xgboost_ray.data_sources.data_source import DataSource, RayFileType
+from xgboost_ray.data_sources.object_store import ObjectStore
+from xgboost_ray.data_sources.pandas import Pandas
+from xgboost_ray.data_sources.numpy import Numpy
+
+
+class Partitioned(DataSource):
+    """Read from distributed data structure implementing __partitioned__.
+
+    __partitioned__ provides meta data about how the data is partitioned and
+    distributed across several compute nodes, making supporting objects them
+    suitable for distributed loading.
+    """
+    supports_central_loading = True
+    supports_distributed_loading = True
+
+    @staticmethod
+    def is_data_type(data: Any,
+                     filetype: Optional[RayFileType] = None) -> bool:
+        return hasattr(data, "__partitioned__")
+
+    @staticmethod
+    def load_data(
+            data: Any,  # __partitioned__ dict
+            ignore: Optional[Sequence[str]] = None,
+            indices = None,  # FIXME Optional[Union[Sequence[int], Sequence[ObjectRef]]] = None,
+            **kwargs) -> pd.DataFrame:
+
+        assert isinstance(data, dict), "Expected __partitioned__ dict"
+        assert len(indices) == 0 or not isinstance(indices[0], int)
```

Comment:
I think a more informative message (like a line above) would be good here

Reply:
Hm, Besides being incorrect (e.g. if indices is None) I guess this was mostly a debug help to understand what's happening. I am not sure I know what the type annotations (as copied from other data sources) allow list entries to be integers and if this can happen at all here. Any thoughts?

Reply:
If this is just a debug thing, then I think we can just remove it

Reply:
I think I understand this now, I will fix the type annotation and remove the assertion.